audio/drivers/coreaudio.cpp

```diff
@@ -0,0 +1,409 @@
+    
+    UInt32 converterQuality = kAudioConverterQuality_Max;
+    status = AudioUnitSetProperty(*audioUnit, kAudioUnitProperty_RenderQuality, kAudioUnitScope_Global, 0, &converterQuality, sizeof(converterQuality));
+    if(status) {
+        qDebug("Failed to set render quality");
+        goto error;
+    }
+    desiredFormat.mSampleRate = getSampleRate(*audioDevice);
+    desiredFormat.mBytesPerPacket = sizeof(float) * 2;
+    desiredFormat.mBytesPerFrame = sizeof(float) * 2;
+    desiredFormat.mChannelsPerFrame = 2;
+    status = AudioUnitSetProperty(*audioUnit, kAudioUnitProperty_StreamFormat, kAudioUnitScope_Input, 0, &desiredFormat, sizeof(desiredFormat));
+    if(status) {
+        qDebug("Failed to set stream format");
+        goto error;
+    }
+    
+    AURenderCallbackStruct renderCallbackData;
+    renderCallbackData.inputProc = CoreAudioIOProc;
+    renderCallbackData.inputProcRefCon = ca;
+    
+    status = AudioUnitSetProperty(*audioUnit, kAudioUnitProperty_SetRenderCallback, kAudioUnitScope_Output, 0, &renderCallbackData, sizeof(renderCallbackData));
+    if(status) {
+        qDebug("Failed to set render callback");
+        goto error;
+    }
+    
+    AudioUnitInitialize(*audioUnit);
+    
+    return true;
+   error:
+    if(*audioUnit) {
+        AudioComponentInstanceDispose(*audioUnit);
+        *audioUnit = NULL;
+    }
+    *audioDevice = kAudioDeviceUnknown;
+    return false;
+}
+
+static OSStatus CoreAudioIOProc( void *inRefCon,
+                               AudioUnitRenderActionFlags *ioActionFlags,
+                               const AudioTimeStamp *inTimeStamp,
+                               UInt32 inBusNumber,
+                               UInt32 inNumberFrames,
+                               AudioBufferList *ioData)
+{
+      CoreAudio* ca = static_cast<CoreAudio*>(inRefCon);
+
+      seq->setInitialMillisecondTimestampWithLatency();
+
+      long bytesPerFrame = sizeof(float) * ioData->mBuffers[0].mNumberChannels;
+
+      float* buffer = static_cast<float*>(ioData->mBuffers[0].mData);
+      UInt64 frames = ioData->mBuffers[0].mDataByteSize / bytesPerFrame;
+
+      seq->process((unsigned)frames, buffer);
+
+      return noErr;
+}
+
+bool CoreAudio::init(bool)
+      {
+	SInt32 major;
+	SInt32 minor;
+	Gestalt(gestaltSystemVersionMajor, &major);
+	Gestalt(gestaltSystemVersionMinor, &minor);
+	
+      initialized = true;
+    
+	// Starting with 10.6 systems, the HAL notification thread is created internally
+	if (major == 10 && minor >= 6) {
+		CFRunLoopRef theRunLoop = NULL;
+		AudioObjectPropertyAddress theAddress = { kAudioHardwarePropertyRunLoop, kAudioObjectPropertyScopeGlobal, kAudioObjectPropertyElementMaster };
+		OSStatus osErr = AudioObjectSetPropertyData (kAudioObjectSystemObject, &theAddress, 0, NULL, sizeof(CFRunLoopRef), &theRunLoop);
+		if (osErr != noErr) {
+                  initialized = false;
+                  return false;
+		}
+	}
```

Comment:
The lowest version of macOS that MuseScore supports is 10.10. For this reason, and because `Gestalt` is deprecated as of macOS 10.8, we can probably skip the version check and make this whole block unconditional.

Reply:
Resolved in badd47